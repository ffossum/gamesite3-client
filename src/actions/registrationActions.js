/* @flow */

const REGISTRATION_REQUEST = 'registration/request';
// const REGISTRATION_SUCCESS = 'registration/success';
const REGISTRATION_FAILURE = 'registration/failure';

export type Registration = {
  username: string,
  email: string,
  password: string,
  repeatPassword: string,
};
type RegistrationRequestAction = {
  type: 'registration/request',
  payload: Registration,
};
function registrationRequest(
  registration: Registration
): RegistrationRequestAction {
  return {
    type: REGISTRATION_REQUEST,
    payload: registration,
  };
}

type RegistrationSuccessAction = {
  type: 'registration/success',
  payload: Registration,
};

type RegistrationFailureAction = {
  type: 'registration/failure',
  payload: Registration,
};
function registrationFailure(
  registration: Registration
): RegistrationFailureAction {
  return {
    type: REGISTRATION_FAILURE,
    payload: registration,
  };
}

export function register(registration: Registration) {
  return (dispatch: Dispatch<*>) => {
    dispatch(registrationRequest(registration));

    fetch('/api/registration', {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      credentials: 'include',
      body: JSON.stringify(registration),
    }).then(response => {
      if (response.ok) {
        location.reload(); // cookie is set, now refresh page
      } else {
        dispatch(registrationFailure(registration));
      }
    });
  };
}

export type RegistrationAction =
  | RegistrationRequestAction
  | RegistrationSuccessAction
  | RegistrationFailureAction;
