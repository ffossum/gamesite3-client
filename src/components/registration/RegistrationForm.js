/* @flow */
import React from 'react';

export default class RegistrationForm extends React.Component {
  props: {
    register: Function,
  };
  handleSubmit: (e: SyntheticInputEvent) => void;
  handleChange: (e: SyntheticInputEvent) => void;
  state: {
    username: string,
    email: string,
    password: string,
    repeatPassword: string,
  };
  constructor() {
    super();

    this.handleSubmit = e => {
      e.preventDefault();
      this.props.register(this.state);
    };

    this.handleChange = e => {
      const { name, value } = e.target;
      this.setState(() => ({
        [name]: value,
      }));
    };
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <div>
          <label htmlFor="username">Username</label>
          <input
            type="text"
            id="username"
            name="username"
            onChange={this.handleChange}
          />
        </div>

        <div>
          <label htmlFor="email">Email</label>
          <input
            type="email"
            id="email"
            name="email"
            onChange={this.handleChange}
          />
        </div>

        <div>
          <label htmlFor="password">Password</label>
          <input
            type="password"
            id="password"
            name="password"
            onChange={this.handleChange}
          />
        </div>

        <div>
          <label htmlFor="password">Repeat password</label>
          <input
            type="password"
            id="repeatPassword"
            name="repeatPassword"
            onChange={this.handleChange}
          />
        </div>

        <button type="submit">Register</button>
      </form>
    );
  }
}
