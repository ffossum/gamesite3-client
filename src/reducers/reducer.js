/* @flow */
import type { Action } from '../actions/actionTypes';

export type State = {};
const initialState = {};
export default function reducer(state?: State = initialState, action: Action) {
  switch (action.type) {
    default:
      return state;
  }
}
