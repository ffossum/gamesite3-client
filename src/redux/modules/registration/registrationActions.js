/* @flow */
export const REGISTRATION_REQUEST = 'registration/request';
export const REGISTRATION_SUCCESS = 'registration/success';
export const REGISTRATION_FAILURE = 'registration/failure';

import type { SessionUser } from '../session/sessionReducer';

export type Registration = {
  username: string,
  email: string,
  password: string,
  repeatPassword: string,
};
export type RegistrationRequestAction = {
  type: 'registration/request',
  payload: Registration,
};
export function registrationRequest(
  registration: Registration
): RegistrationRequestAction {
  return {
    type: REGISTRATION_REQUEST,
    payload: registration,
  };
}

type RegistrationSuccessAction = {
  type: 'registration/success',
  payload: SessionUser,
};
export function registrationSuccess(
  user: SessionUser
): RegistrationSuccessAction {
  return {
    type: REGISTRATION_SUCCESS,
    payload: user,
  };
}

type RegistrationFailureAction = {
  type: 'registration/failure',
};
export function registrationFailure(): RegistrationFailureAction {
  return {
    type: REGISTRATION_FAILURE,
  };
}

export type RegistrationAction =
  | RegistrationRequestAction
  | RegistrationSuccessAction
  | RegistrationFailureAction;
