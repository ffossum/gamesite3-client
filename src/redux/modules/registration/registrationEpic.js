/* @flow */
import { ajax } from 'rxjs/observable/dom/ajax';
import {
  REGISTRATION_REQUEST,
  registrationSuccess,
  registrationFailure,
} from './registrationActions';
import type { RegistrationRequestAction } from './registrationActions';

const registrationEpic = (action$: ActionsObservable<*>) =>
  action$
    .ofType(REGISTRATION_REQUEST)
    .mergeMap((action: RegistrationRequestAction) =>
      ajax({
        url: '/api/registration',
        method: 'POST',
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(action.payload),
      })
        .mapTo(registrationSuccess())
        .catch(() => [registrationFailure(action.payload)])
    );

export default registrationEpic;
