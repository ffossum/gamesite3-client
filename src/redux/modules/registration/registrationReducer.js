/* @flow */
import type { Action } from '../../actions';
import {
  REGISTRATION_FAILURE,
  REGISTRATION_REQUEST,
  REGISTRATION_SUCCESS,
} from './registrationActions';

export type RegistrationState = {
  loading: boolean,
};
const initialState = {
  loading: false,
};
export default function reducer(
  state?: RegistrationState = initialState,
  action: Action
) {
  switch (action.type) {
    case REGISTRATION_REQUEST: {
      return {
        ...state,
        loading: true,
      };
    }

    case REGISTRATION_SUCCESS:
    case REGISTRATION_FAILURE: {
      return {
        ...state,
        loading: false,
      };
    }

    default:
      return state;
  }
}
