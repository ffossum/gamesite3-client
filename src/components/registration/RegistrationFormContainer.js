/* @flow */
import { connect } from 'react-redux';
import RegistrationForm from './RegistrationForm';
import { register } from '../../actions/registrationActions';

export default connect(
  () => ({}),
  (dispatch: Dispatch<*>) => ({
    register(...args) {
      dispatch(register(...args));
    },
  })
)(RegistrationForm);
