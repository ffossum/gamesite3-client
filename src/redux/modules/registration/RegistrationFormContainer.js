/* @flow */
import { connect } from 'react-redux';
import RegistrationForm from '../../../components/registration/RegistrationForm';
import { registrationRequest } from './registrationActions';

import type { State } from '../root';

export default connect(
  (state: State) => ({
    loading: state.registration.loading,
  }),
  (dispatch: Dispatch<*>) => ({
    register(...args) {
      dispatch(registrationRequest(...args));
    },
  }),
)(RegistrationForm);
