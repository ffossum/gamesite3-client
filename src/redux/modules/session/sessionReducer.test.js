/* @flow */
/* eslint-env jest */
import reducer from './sessionReducer';
import { registrationSuccess } from '../registration/registrationActions';

describe('session reducer', () => {
  const initialState = reducer(undefined, { type: '@@INIT' });
  test('initial state', () => {
    expect(initialState).toMatchSnapshot();
  });
  test('contains user data after successful registration', () => {
    const action = registrationSuccess({
      username: 'bob',
      id: 'userid',
    });

    const state = reducer(initialState, action);
    expect(state).toEqual({
      user: {
        username: 'bob',
        id: 'userid',
      },
    });
  });
});
