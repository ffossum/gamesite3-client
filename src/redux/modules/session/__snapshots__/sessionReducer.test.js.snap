// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`session reducer initial state 1`] = `
Object {
  "user": null,
}
`;
